src/blueapi/config.py

```diff
@@ -117,54 +133,47 @@ def __init__(self, schema: type[C]) -> None:
 
     def use_values(self, values: Mapping[str, Any]) -> None:
         """
-        Use all values provided in the config, override any defaults
-        and values set by previous calls into this class.
-
-        Args:
-            values (Mapping[str, Any]): Dictionary of override values,
-                                        does not need to be exhaustive
-                                        if defaults provided.
+        Use all values provided in the config, override any defaults.
         """
-
-        def recursively_update_map(old: dict[str, Any], new: Mapping[str, Any]) -> None:
-            for key in new:
-                if (
-                    key in old
-                    and isinstance(old[key], dict)
-                    and isinstance(new[key], dict)
-                ):
-                    recursively_update_map(old[key], new[key])
-                else:
-                    old[key] = new[key]
-
-        recursively_update_map(self._values, values)
+        self._values = _recursively_updated_map(self._values, values)
 
     def use_values_from_yaml(self, path: Path) -> None:
         """
-        Use all values provided in a YAML/JSON file in the
-        config, override any defaults and values set by
-        previous calls into this class.
-
-        Args:
-            path (Path): Path to YAML/JSON file
+        Use values from a YAML/JSON file, overriding previous values.
         """
-
         with path.open("r") as stream:
             values = yaml.load(stream, yaml.Loader)
         self.use_values(values)
 
-    def load(self) -> C:
+    def use_values_from_env(self) -> None:
         """
-        Finalize and load the config as an instance of the `schema`
-        dataclass.
-
-        Returns:
-            C: Dataclass instance holding config
+        Load values from environment variables with a given prefix.
         """
 
+        env_values = {}
+        for key, value in os.environ.items():
+            if key.startswith(ENV_PREFIX) and value is not None:
+                # Convert key to a config path-like structure
+                config_key = key.removeprefix(ENV_PREFIX).lower()
+                li: list[str] = config_key.split(".")
+                local_value = env_values
+                while len(li) > 1:
+                    key = li.pop(0)
+                    local_value = local_value.setdefault(key, {})
+                local_value.setdefault(li[0], value)
```

Comment:
```suggestion
                *path, key = config.split('_') # See ENV_PREFIX
                local_value = env_values
                for entry in path:
                    local_value = local_value.setdefault(entry, {})
                local_value.setdefault(key, value)
```

This should probably also handle the error case where keys are incompatible

```
BLUEAPI_config_foo=3 BLUEAPI_config_foo_bar=4 blueapi serve
```
currently (assuming the `.`/`_` problems are fixed) fails with
```
AttributeError: 'str' object has no attribute 'setdefault'
```
which isn't very informative

Reply:
good catch